### PR TITLE
[Merged by Bors] - feat: meta code for the category_theory port

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

File: Mathlib/Lean/Meta.lean [hunk 1]
```diff
@@ -50,3 +50,22 @@ def _root_.Lean.MVarId.existsi (mvar : MVarId) (es : List Expr) : MetaM MVarId :
...
     mvar
+
+/--
+Given a monadic function `deduce` that takes a type and a term of that type and produces a new term,
+lifts this to the monadic function that opens a `∀` telescope, applies `deduce` to the body,
+and then builds the lambda telescope term for the new term.
+-/
+def mapForallTelescope' (deduce : Expr → Expr → MetaM Expr) (forallTerm : Expr) : MetaM Expr := do
+  forallTelescope (← Meta.inferType forallTerm) fun xs ty => do
+    Meta.mkLambdaFVars xs (← deduce ty (mkAppN forallTerm xs))
```
Comment (semorrison) [inline Mathlib/Lean/Meta.lean:61]:
```suggestion
Given a monadic function `F` that takes a type and a term of that type and produces a new term,
lifts this to the monadic function that opens a `∀` telescope, applies `F` to the body,
and then builds the lambda telescope term for the new term.
-/
def mapForallTelescope' (F : Expr → Expr → MetaM Expr) (forallTerm : Expr) : MetaM Expr := do
  forallTelescope (← Meta.inferType forallTerm) fun xs ty => do
    Meta.mkLambdaFVars xs (← F ty (mkAppN forallTerm xs))
```

File: Mathlib/Lean/Meta.lean [hunk 2]
```diff
@@ -50,3 +50,22 @@ def _root_.Lean.MVarId.existsi (mvar : MVarId) (es : List Expr) : MetaM MVarId :
...
+
+/--
+Given a monadic function `deduce` that takes a term and produces a new term,
+lifts this to the monadic function that opens a `∀` telescope, applies `deduce` to the body,
+and then builds the lambda telescope term for the new term.
+-/
+def mapForallTelescope (deduce : Expr → MetaM Expr) (forallTerm : Expr) : MetaM Expr := do
+  mapForallTelescope' (fun _ e => deduce e) forallTerm
```
Comment (semorrison) [inline Mathlib/Lean/Meta.lean:69]:
```suggestion
/--
Given a monadic function `F` that takes a term and produces a new term,
lifts this to the monadic function that opens a `∀` telescope, applies `F` to the body,
and then builds the lambda telescope term for the new term.
-/
def mapForallTelescope (F : Expr → MetaM Expr) (forallTerm : Expr) : MetaM Expr := do
  mapForallTelescope' (fun _ e => F e) forallTerm
```

File: Mathlib/Lean/Meta/Simp.lean [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  lemmas.foldlM (·.addConst ·) (← simpOnlyBuiltins.foldlM (·.addConst ·) {})
+
+/-- Simplify an expression using only a list of lemmas specified by name. -/
+def simpOnlyNames (lemmas : List Name) (e : Expr) : MetaM Simp.Result := do
+  (·.1) <$> simp e
+    { simpTheorems := #[← simpTheoremsOfNames lemmas], congrTheorems := ← getSimpCongrTheorems }
+
+/--
+Given a simplifier `S : Expr → MetaM Simp.Result`,
+and an expresssion `e : Expr`, run `S` on the type of `e`, and then
```
Comment (fpvandoorn) [inline Mathlib/Lean/Meta/Simp.lean:27]:
```suggestion
and an expression `e : Expr`, run `S` on the type of `e`, and then
```

File: Mathlib/Lean/Meta/Simp.lean [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+open Lean Elab.Tactic
+
+namespace Lean.Meta
+
+/-- Construct a `SimpTheorems` from a list of names. (i.e. as with `simp only`). -/
+def simpTheoremsOfNames (lemmas : List Name) : MetaM SimpTheorems := do
+  lemmas.foldlM (·.addConst ·) (← simpOnlyBuiltins.foldlM (·.addConst ·) {})
+
+/-- Simplify an expression using only a list of lemmas specified by name. -/
+def simpOnlyNames (lemmas : List Name) (e : Expr) : MetaM Simp.Result := do
```
Comment (fpvandoorn) [inline Mathlib/Lean/Meta/Simp.lean:21]: We need to write a `mkSimpContext` with all the bells an whistles of `simp` (and then maybe PR that to core).
Another partial implementation is in `Tactic.Simps.Basic` (that should be moved to this file).

**Review by fpvandoorn — APPROVED** [review]
bors d+

File: Mathlib/Lean/Meta/Simp.lean (hunk 4 above)
Comment (semorrison) [inline Mathlib/Lean/Meta/Simp.lean:21]:
```suggestion
-- TODO We need to write a `mkSimpContext` in `MetaM`
-- that supports all the bells and whistles in `simp`.
-- It should generalize this, and another partial implementation in `Tactic.Simps.Basic`.
def simpOnlyNames (lemmas : List Name) (e : Expr) : MetaM Simp.Result := do
```

Comment (semorrison) [reply Mathlib/Lean/Meta/Simp.lean:21]: Thanks. I've left a comment pointing to that other partial implementation.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**